runtime/src/util/mapmap.rs

```diff
@@ -0,0 +1,111 @@
+pub struct MapMap<'a, BS, V, K1, K2> {
+    outer: Map<'a, BS, Cid>,
+    inner_bitwidth: u32,
+    // cache all inner maps loaded since last load/flush
+    // get/put/remove operations load the inner map into the cache first and modify in memory
+    // flush writes all inner maps in the cache to the outer map before flushing the outer map
+    cache: BTreeMap<Vec<u8>, Map<'a, BS, V>>,
+    key_types: PhantomData<(K1, K2)>,
+}
+impl<'a, BS, V, K1, K2> MapMap<'a, BS, V, K1, K2>
+where
+    BS: Blockstore,
+    V: Serialize + DeserializeOwned + Clone + std::cmp::PartialEq,
+    K1: Keyer + std::fmt::Debug + std::fmt::Display,
+    K2: Keyer + std::fmt::Debug + std::fmt::Display,
+{
+    pub fn new(bs: &'a BS, outer_bitwidth: u32, inner_bitwidth: u32) -> Self {
+        MapMap {
+            outer: make_empty_map(bs, outer_bitwidth),
+            inner_bitwidth,
+            cache: BTreeMap::<Vec<u8>, Map<BS, V>>::new(),
+            key_types: PhantomData,
+        }
+    }
+
+    pub fn from_root(
+        bs: &'a BS,
+        cid: &Cid,
+        outer_bitwidth: u32,
+        inner_bitwidth: u32,
+    ) -> Result<Self, Error> {
+        Ok(MapMap {
+            outer: make_map_with_root_and_bitwidth(cid, bs, outer_bitwidth)?,
+            inner_bitwidth,
+            cache: BTreeMap::<Vec<u8>, Map<BS, V>>::new(),
+            key_types: PhantomData,
+        })
+    }
+
+    pub fn flush(&mut self) -> Result<Cid, Error> {
+        for (k, in_map) in self.cache.iter_mut() {
+            if in_map.is_empty() {
+                self.outer.delete(&BytesKey(k.to_vec()))?;
+            } else {
+                let new_in_root = in_map.flush()?;
+                self.outer.set(BytesKey(k.to_vec()), new_in_root)?;
+            }
+        }
+        self.outer.flush()
+    }
+
+    fn load_inner_map(&mut self, k: K1) -> Result<&mut Map<'a, BS, V>, Error> {
+        let in_map_thunk = || -> Result<Map<BS, V>, Error> {
+            // lazy to avoid ipld operations in case of cache hit
+            match self.outer.get(&k.key())? {
+                Some(root) => Ok(make_map_with_root_and_bitwidth::<BS, V>(
+                    root,
+                    *self.outer.store(),
+                    self.inner_bitwidth,
+                )?),
+                None => Ok(make_empty_map(*self.outer.store(), self.inner_bitwidth)),
+            }
+        };
+        let raw_k = k.key().0;
+        match self.cache.entry(raw_k) {
+            Occupied(entry) => Ok(entry.into_mut()),
+            Vacant(entry) => Ok(entry.insert(in_map_thunk()?)),
+        }
+    }
+
+    // memreplace -- lets you swap two values without triggering borrow checker
+    // cloning something somewhere, doing this insert on some cloned version of the cache
+    pub fn get(&mut self, outside_k: K1, inside_k: K2) -> Result<Option<&V>, Error> {
+        let in_map = self.load_inner_map(outside_k)?;
+        in_map.get(&inside_k.key())
```

Comment:
It's a bit unfortunate that this will create an empty inner map if K1 didn't exist, only to fail to find K2 in it. Ok to defer improving this to later (file an issue)

Reply:
Nice catch should be fixed